Source/InvocationCollection.cs

```diff
@@ -47,20 +47,20 @@ namespace Moq
 {
 	internal sealed class InvocationCollection : IReadOnlyList<IReadOnlyInvocation>
 	{
-		private List<Invocation> invocations;
+		private Invocation[] invocations;
```

Comment:
I suppose I might be missing something obvious here, but I'd really like to avoid us reinventing the wheel and programming another version of `List<>`. Why is this necessary?

Reply:
If we continue to use `List<T>` you also have to use `List<T>.Enumerator` which isn't going to behave the way we want in concurrent-execution scenarios. Arrays, being primitives, have a much simpler access model, so we can know and take advantage of what operations are safe to do concurrently and which are not.

Reply:
Wouldn't we only end up with `List<T>.Enumerator` if we used LINQ or a `foreach` loop? What if we went for a simple `for` loop that accesses the list by index?

Reply:
I did consider it, but it has the same problem. When you call `List<T>.Add(T)`, it's not concurrency-safe with its indexer. A path exists where you wind up in an unknown state, like if the underlying array has to be resized. We could try to work out how the implementation of `List<T>` works and try to handle these edge-cases, but I think it's considerably easier to just use an array. The only method with any real complexity is our `Add(Invocation)` and most of that is delegated to the fantastic `Array.Resize` method.

Reply:
I can't see the problem yet. Due to the locks in `InvocationCollection`, only one thread can ever access the list at a time. If an `Add` causes a resize, then that resize happens inside the `lock`. The indexer access also happens inside a `lock`, so while that happens, noone can modify the list.

Reply:
Sorry if the following might be a little naïve&mdash;not sure, a single `lock` around the whole method body might be sufficient, not sure how the C# compiler transforms `lock` blocks when `yield return` gets involved:

```csharp
IEnumerator<IReadOnlyInvocation> GetEnumerator()
{
    List<Invocation> invocations;
    int count;
    lock (this.invocationsLock)
    {
        invocations = this.invocations;
        count = invocations.Count;
    }
    for (int i = 0; i < count; ++i)
    {
        lock (this.invocationsLock)
        {
            yield return invocations[i];
        }
    }
}
```

Reply:
A `lock` around the `for` loop would actually make it so the enumerator would acquire the collection until all elements had been iterated over. What you have should acquire a lock to the collection every time it yields an item, then release the lock.

I do like how much less code there is here. I also like that my implementation only acquires a lock once, versus once-per-item + 1 times. I could go either way.

Reply:
Perhaps there is a middle way. I feel it should be possible to let the compiler generate the enumerator class, which would help reduce the amount of additional code. I'll look into this more closely a little later.

Reply:
Added an implementation using `yield return`. It should be functionally identical to the original, except it doesn't support `IEnumerator.Reset()`.